src/els_code_action_provider.erl

```diff
@@ -0,0 +1,128 @@
+%% els_provider functions
+%%==============================================================================
+
+-spec is_enabled() -> boolean().
+is_enabled() -> true.
+
+%% /**
+%%  * The server provides code actions. The `CodeActionOptions` return type is
+%%  * only valid if the client signals code action literal support via the
+%%  * property `textDocument.codeAction.codeActionLiteralSupport`.
+%%  */
+%% codeActionProvider?: boolean | CodeActionOptions;
+
+-spec options() -> boolean() | map().
+options() ->
+  Capabilities = els_config:get(capabilities),
+  lager:info("code_actions: [Capabilities=~p]", [Capabilities]),
+  true.
+
+-spec handle_request(any(), els_provider:state()) ->
+  {any(), els_provider:state()}.
+handle_request({document_codeaction, Params}, State) ->
+  #{ <<"textDocument">> := #{ <<"uri">> := Uri}
+   , <<"range">>   := RangeLSP
+   , <<"context">> := Context } = Params,
+  Result = code_actions(Uri, RangeLSP, Context),
+  {Result, State}.
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+
+%% result: (Command | CodeAction)[] | null where CodeAction is defined
+%% as follows:
+-spec code_actions(uri(), range(), code_action_context()) -> [map()].
+code_actions(Uri, Range, Context) ->
+  lager:info("code_actions: [Context=~p]", [Context]),
+  #{ <<"start">> := #{ <<"character">> := _StartCol
+                     , <<"line">>      := StartLine }
+   , <<"end">> := _End
+   } = Range,
+  #{ <<"diagnostics">> := Diagnostics } = Context,
+  %% TODO: process optional "only" field in Context
+  Actions0 = [ make_code_action(Uri, C) || C <- Diagnostics],
+
+  Actions = case application:get_env(erlang_ls, test_code_action) of
+    {ok, true} -> [ add_lines_action( Uri
+                                    , <<"Add TODO comment">>
+                                    , <<"%% TODO: something\n">>
+                                    , StartLine)]
+                   ++ Actions0;
+    _ -> Actions0
+  end,
+  lager:info("code_actions: [Actions=~p]", [Actions]),
+  Actions.
+
+
+%% TODO: either provide a literal workspaceEdit, or a command, based
+%% on the `textDocument.codeAction.codeActionLiteralSupport` client
+%% capability.
+-spec add_lines_action(uri(), binary(), binary(), number()) -> map().
+add_lines_action(Uri, Title, Lines, Before) ->
+   #{ title => Title
+    , command =>
+          els_utils:make_command( Title
+                                , <<"add-lines">>
+                                , [#{ uri => Uri
+                                    , lines => Lines
+                                    , before => Before }])
+    }.
+
+-spec make_code_action(uri(), diagnostic()) -> [code_action()].
+make_code_action(Uri, #{ <<"message">> := Message
+                       , <<"range">> := Range } = _Diagnostic) ->
```

Comment:
Using a regexp on the diagnostic message seems a fragile approach, since messages could vary between tools versions.
Diagnostics have a 'code' field which identify the diagnostic type. Maybe we could leverage that to have an exact match?

Reply:
I understand the `code` field to be used as a reference back to the compiler spec in terms of the kind of error found, which means it would be a stretch to pack this info into it.

The fundamental problem is that we are working with what the compiler emits as a warning or error.  So we have to parse the error message some time, either when we generate the diagnostic, or when we generate the code action. The best approach is to modify the compiler to provide a structured diagnostic that we can use directly.

cc @michalmuskala  @ilya-klyuchnikov

Reply:
Sorry, I didn't see this comment before.

In general, it's possible to get structured output from the Erlang compiler. Calling the `compile:file/2` function with the `return` option, causes it to produce something like:
```erl
{error, 
	[{"test.erl", [{3,erl_lint,{undefined_function,{foo,1}}}]}],
	[]}
```
The text-based message might be recovered by calling, in this example, `erl_lint:format_error({undefined_function,{foo,1}})`. As far as I understand, this structured message should be treated as opaque and there are no guarantees about its stability, but in practice it seems to be stable enough. The Elixir compiler was matching on messages like this to emit proper error messages related to the original Elixir source before rewriting the linter pass and still matches on some of the messages coming from [deeper compiler passes](https://github.com/elixir-lang/elixir/blob/033c7387e25aa31fb4d1bf52d2b5cef6519e9df7/lib/elixir/src/elixir_erl_compiler.erl#L87-L150).

Reply:
@michalmuskala @alanz Compiler diagnostics use exactly this mechanism. See https://github.com/erlang-ls/erlang_ls/blob/634cc2ad63f8b1c5675737ea9b62764fe912630c/src/els_compiler_diagnostics.erl#L74 and https://github.com/erlang-ls/erlang_ls/blob/634cc2ad63f8b1c5675737ea9b62764fe912630c/src/els_compiler_diagnostics.erl#L137 (since the callback module can be something other than `erl_lint`).

Reply:
Ok, so we are back to the problem of how to associate the diagnostic with the original compiler warning/error.  There is no "additional info" field in the `Diagnostic` type, and I think using the `code` field is inappropriate.

```typescript
/**
  * The diagnostic's code, which might appear in the user interface.
  */
	code?: number | string;
```

Perhaps we should store it, against a key made up of the diagnostic itself, or some key subset of its fields?

Reply:
FYI: https://github.com/microsoft/language-server-protocol/issues/887

Reply:
Even if we find a way to reference an existing diagnostic, this would imply we need to store diagnostics on the server side, which would require many changes.
In 3.15 they added some (very limited) "tags" to the Diagnostic. That should probably be extended to include generic metadata. I agree `code` does not sound correct, but it feels like it's the only thing we have right now. Is there any documentation explaining the idea behind the `code` field? It's still not clear to me what it's supposed to represent.

For the time being, considering we don't have any real code action, what do you think about splitting this PR into parts (one for the code lenses support, one for the execute command), so we can easily merge those and hold on the code action part?

Reply:
We don't have a code lens use case now either.

Perhaps I should split it into adding the machinery without the test, and keeping the test with its config in a separate branch as a model for when we do actually implement one.

I am not all that happy to have a feature in place that is noc "claimed" by a test though, the danger exists that it can rot.